# Tutorial 1: Python Basics for Data Analysis

**CS Statistics Course | Week 1-2**

---

## Welcome to Data Analysis!

In Lectures 1-2, you learned what data science is and why it matters. You saw that data scientists work with:
- **Categorical data** (like programming languages, company sizes)
- **Numerical data** (like salaries, years of experience)
- **Data summaries** (mean, median, standard deviation, quantiles)
- **Data visualization** (histograms, bar charts, scatter plots)

In this tutorial, you'll **get hands-on** with the Python tools that let you do this analysis yourself. By the end of 2 hours, you'll be able to load real data, explore it, summarize it, and visualize it — exactly like a data scientist.

### The Case Study: Analyzing Software Developer Survey Data

Imagine you work at a tech recruitment company and want to understand the software developer job market. You have survey data from 500 developers around the world with information about:
- Their primary programming language
- Years of experience
- Annual salary
- Job satisfaction
- Company size (startup, small, medium, large, enterprise)
- Remote work status (yes, no, hybrid)
- Weekly hours worked
- Number of projects
- Country

**Your goal:** Use Python to answer real questions like:
- What's the average salary for developers with 5 years of experience?
- Do developers who work remotely have higher job satisfaction?
- How do salaries differ across programming languages?
- What's the relationship between experience and salary?

Let's get started!


## Setup: Import the Libraries

We'll use four key libraries for data analysis:
- **pandas** ([documentation](https://pandas.pydata.org/docs/)) — for loading and manipulating data
- **numpy** ([documentation](https://numpy.org/doc/)) — for numerical computations
- **matplotlib** ([documentation](https://matplotlib.org/)) — for basic plotting
- **seaborn** ([documentation](https://seaborn.pydata.org/)) — for statistical visualization


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set up plotting style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)

---

## Section 1: Loading and Exploring Data (~15 min)

### Why This Matters (From Lecture)
Before you can analyze data, you need to understand what you're working with. In lecture, you learned that every dataset has:
- **Rows** (observations/records) — each row is one developer
- **Columns** (features/variables) — each column is a characteristic like salary or experience

Let's load the developer survey data and explore it.


### Loading Data with `pd.read_csv()`

The most common way to load data into Python is from a CSV (comma-separated values) file.

Reference: [pd.read_csv documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)


In [2]:
# Load the developer survey dataset
df = pd.read_csv('developer_survey.csv')
print("Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows, {df.shape[1]} columns")

Dataset loaded successfully!
Shape: 500 rows, 10 columns


### Exploring the Dataset

Use these methods to understand your data:
- `df.head()` — see the first few rows
- `df.shape` — get the dimensions (rows, columns)
- `df.dtypes` — check data types (int, float, object/string)
- `df.describe()` — get summary statistics
- `df.info()` — get detailed information including missing values


In [3]:
# See the first 5 rows
df.head()

,developer_id,primary_language,years_experience,annual_salary,job_satisfaction,company_size,remote_work,weekly_hours,num_projects,country
0,362,JavaScript,5,47909.0,6.0,medium,yes,28.0,2,India
1,74,C++,8,62777.0,3.0,large,hybrid,26.0,1,India
2,375,Python,18,66011.0,5.0,medium,no,NaN,5,India
3,156,Python,19,103941.0,6.0,large,no,49.0,9,UK
4,105,Rust,15,103425.0,5.0,medium,yes,59.0,4,UK


In [ ]:
# Check the shape (rows, columns)
print(f"Dataset shape: {df.shape}")
print(f"Number of developers: {df.shape[0]}")
print(f"Number of features per developer: {df.shape[1]}")

In [ ]:
# Check data types
print("Data types:")
print(df.dtypes)
print("\nNote:")
print("- int64: integers (whole numbers)")
print("- float64: floating-point numbers (with decimals) — may contain NaN values")
print("- object: strings/text data")

In [ ]:
# Get summary statistics (matches what you learned in lecture!)
df.describe()

In [ ]:
# Get detailed info including missing values
df.info()

### Selecting Columns

To work with a specific column, you have two options:
- `df["column_name"]` (always works)
- `df.column_name` (works if column name has no spaces)

Reference: [Indexing and selecting data](https://pandas.pydata.org/docs/user_guide/indexing.html)


In [ ]:
# Select a single column — returns a Series
languages = df['primary_language']
print(f"Type: {type(languages)}")
print(f"\nFirst 5 values:")
print(languages.head())

In [ ]:
# Select multiple columns — returns a DataFrame
subset = df[['primary_language', 'years_experience', 'annual_salary']]
print(f"Type: {type(subset)}")
print(f"\nShape: {subset.shape}")
print(f"\nFirst few rows:")
print(subset.head())

### Selecting Rows

To select specific rows, use:
- `df.loc[row_label]` — select by label (row index)
- `df.iloc[row_position]` — select by integer position

Reference: [loc and iloc documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html)


In [ ]:
# Select the first developer (row 0)
first_dev = df.iloc[0]
print("First developer's data:")
print(first_dev)
print(f"\nType: {type(first_dev)}")

In [ ]:
# Select the first 10 rows
first_10 = df.iloc[0:10]
print(f"First 10 developers (shape: {first_10.shape}):")
print(first_10)

In [ ]:
# Select a specific cell: developer at row 5, 'annual_salary' column
salary_row5 = df.loc[5, 'annual_salary']
print(f"Developer at index 5's salary: ${salary_row5:,.0f}")

### EXERCISES: Section 1

**Exercise 1.1:** How many columns are in the dataset, and what is the name of the 5th column?


In [ ]:
# Your code here

**Exercise 1.2:** Extract the `years_experience` column and show the first 10 values.


In [ ]:
# Your code here

**Exercise 1.3:** Select the rows from index 100 to 105 and show the `developer_id`, `primary_language`, and `annual_salary` columns.


In [ ]:
# Your code here

---

## Section 2: Filtering and Transforming Data (~20 min)

### Why This Matters (From Lecture)
Real data is messy and includes everything. To answer specific questions, you need to:
- **Filter** the data to focus on subsets (e.g., "only developers with 10+ years of experience")
- **Transform** the data (e.g., convert salary to a different unit, standardize values)
- **Handle missing data** (some surveys have incomplete responses)

Let's master these techniques.


### Boolean Masking: Filtering with Conditions

Use boolean conditions to select rows that meet specific criteria.

Reference: [Boolean indexing](https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing)


In [ ]:
# Create a boolean mask for developers with 10+ years of experience
experienced = df['years_experience'] >= 10
print(f"Type of mask: {type(experienced)}")
print(f"\nFirst 10 values (True = meets condition, False = doesn't):")
print(experienced.head(10))
print(f"\nNumber of developers with 10+ years: {experienced.sum()}")

In [ ]:
# Use the mask to filter the dataframe
experienced_devs = df[experienced]
print(f"Shape of filtered data: {experienced_devs.shape}")
print(f"\nFirst few experienced developers:")
print(experienced_devs.head())

In [ ]:
# You can combine filter and selection in one line
result = df[df['years_experience'] >= 10][['primary_language', 'years_experience', 'annual_salary']]
print(f"Developers with 10+ years (showing languages, experience, salary):")
print(result.head())

### Combining Multiple Conditions

Use `&` (AND), `|` (OR), and `~` (NOT) to combine conditions.

**Important:** Always wrap conditions in parentheses!


In [ ]:
# AND: Developers with 10+ years AND working in Python
experienced_python = df[(df['years_experience'] >= 10) & (df['primary_language'] == 'Python')]
print(f"Experienced Python developers: {len(experienced_python)}")
print(f"\nAverage salary: ${experienced_python['annual_salary'].mean():,.0f}")
print(f"\nFirst few:")
print(experienced_python[['primary_language', 'years_experience', 'annual_salary']].head())

In [ ]:
# OR: Developers using Rust OR Go
rust_or_go = df[(df['primary_language'] == 'Rust') | (df['primary_language'] == 'Go')]
print(f"Developers using Rust or Go: {len(rust_or_go)}")
print(f"\nLanguage breakdown:")
print(rust_or_go['primary_language'].value_counts())# Your code here
# OR: Developers using Rust OR Go

In [ ]:
# NOT: Developers NOT working remotely
not_remote = df[~(df['remote_work'] == 'yes')]
print(f"Developers not working remotely: {len(not_remote)}")
print(f"\nRemote work breakdown:")
print(not_remote['remote_work'].value_counts())

### Transforming Data with `.apply()`

The `.apply()` method lets you apply a function to every value in a column.

Reference: [apply documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.apply.html)


In [ ]:
# Example: Convert salary from dollars to thousands of dollars
salary_in_thousands = df['annual_salary'].apply(lambda x: x / 1000 if pd.notna(x) else np.nan)
print(f"Original salary (first 5): {df['annual_salary'].head().values}")
print(f"Converted (first 5): {salary_in_thousands.head().values}")

In [ ]:
# Define a function and use it with apply
def salary_bracket(salary):
    """Assign salary to a bracket."""
    if pd.isna(salary):
        return 'Unknown'
    elif salary < 60000:
        return 'Junior'
    elif salary < 100000:
        return 'Mid-level'
    else:
        return 'Senior'

df['salary_bracket'] = df['annual_salary'].apply(salary_bracket)
print(f"Salary bracket distribution:")
print(df['salary_bracket'].value_counts())
print(f"\nFirst few rows with new column:")
print(df[['annual_salary', 'salary_bracket']].head(10))

### Handling Missing Values

Real data often has missing values. Learn to detect and handle them.

Reference: [Working with missing data](https://pandas.pydata.org/docs/user_guide/missing_data.html)


In [ ]:
# Check for missing values
print("Missing values per column:")
missing_counts = df.isna().sum()
print(missing_counts[missing_counts > 0])
print(f"\nPercentage of data that's missing:")
missing_pct = (df.isna().sum() / len(df) * 100)
print(missing_pct[missing_pct > 0])

In [ ]:
# Filter to only rows with non-null salary
complete_salary = df[df['annual_salary'].notna()]
print(f"Rows with valid salary data: {len(complete_salary)} out of {len(df)}")

# Drop all rows with ANY missing values
complete = df.dropna()
print(f"Rows with complete data (no missing): {len(complete)} out of {len(df)}")

In [ ]:
# Fill missing values with a specific value (e.g., median)
df_filled = df.copy()
df_filled.fillna({'annual_salary': df['annual_salary'].median()}, inplace=True)
df_filled.fillna({'job_satisfaction' : df['job_satisfaction'].median()}, inplace=True)
df_filled.fillna({'weekly_hours': df['weekly_hours'].median()}, inplace=True)

print(f"Missing values after filling:")
print(df_filled.isna().sum().sum())  # Total missing values

### EXERCISES: Section 2

**Exercise 2.1:** Filter the dataset to show only developers from the US who work remotely (full-time remote, not hybrid).


In [ ]:
# Your code here

**Exercise 2.2:** Create a new column `experience_level` that assigns developers to "Beginner" (< 5 years), "Intermediate" (5-15 years), or "Expert" (15+ years).


In [ ]:
# Your code here

**Exercise 2.3:** How many developers have missing data for job_satisfaction? What percentage of the dataset is this?


In [ ]:
# Your code here

---

## Section 3: Data Summaries (~20 min)

### Why This Matters (From Lecture)
In lecture, you learned about **summary statistics**: mean, median, standard deviation, min, max, quantiles. These help you understand the "typical" value in your data and how spread out it is.

Now you'll compute them yourself!


### Basic Summary Statistics

Pandas provides many methods to summarize numerical data.

Reference: [Descriptive statistics](https://pandas.pydata.org/docs/user_guide/basics.html#descriptive-statistics)


In [ ]:
# Get individual statistics
salary_data = df['annual_salary'].dropna()  # Remove NaN for these calculations

print("Salary Statistics:")
print(f"  Mean: ${salary_data.mean():,.0f}")
print(f"  Median: ${salary_data.median():,.0f}")
print(f"  Std Dev: ${salary_data.std():,.0f}")
print(f"  Min: ${salary_data.min():,.0f}")
print(f"  Max: ${salary_data.max():,.0f}")
print(f"  25th percentile (Q1): ${salary_data.quantile(0.25):,.0f}")
print(f"  50th percentile (Q2/Median): ${salary_data.quantile(0.50):,.0f}")
print(f"  75th percentile (Q3): ${salary_data.quantile(0.75):,.0f}")

In [ ]:
# The .describe() method gives you most of these at once
df['annual_salary'].describe()

In [ ]:
# Get describe for multiple columns
df[['years_experience', 'annual_salary', 'job_satisfaction']].describe()

### Grouped Summaries with `.groupby()`

One of the most powerful tools in pandas! Group data by a categorical variable, then compute statistics for each group.

Reference: [groupby documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html)


In [ ]:
# Group by primary_language and get mean salary for each
salary_by_language = df.groupby('primary_language')['annual_salary'].mean().sort_values(ascending=False)
print("Average salary by programming language:")
for lang, salary in salary_by_language.items():
    print(f"  {lang}: ${salary:,.0f}")

In [ ]:
# Get multiple statistics at once
language_stats = df.groupby('primary_language')['annual_salary'].agg([
    ('count', 'count'),
    ('mean', 'mean'),
    ('median', 'median'),
    ('std', 'std'),
    ('min', 'min'),
    ('max', 'max')
])
print("\nDetailed salary statistics by language:")
print(language_stats.round(0))

In [ ]:
# Group by company size
salary_by_company = df.groupby('company_size')['annual_salary'].agg(['count', 'mean', 'median'])
print("\nSalary statistics by company size:")
print(salary_by_company.round(0))

In [ ]:
# Group by remote work status
remote_stats = df.groupby('remote_work')[['annual_salary', 'job_satisfaction']].agg(['mean', 'median'])
print("\nSalary and satisfaction by remote work status:")
print(remote_stats.round(0))

### Categorical Data: `.value_counts()`

For categorical columns, use `.value_counts()` to see the frequency of each category.


In [ ]:
# Count how many developers use each language
language_counts = df['primary_language'].value_counts()
print("Number of developers by language:")
print(language_counts)
print(f"\nTotal: {language_counts.sum()}")

In [ ]:
# Get percentages
language_pct = df['primary_language'].value_counts(normalize=True) * 100
print("Percentage of developers by language:")
for lang, pct in language_pct.items():
    print(f"  {lang}: {pct:.1f}%")

In [4]:
# Cross-tabulation: see the breakdown of two categorical variables
# Note: this is just a contingency table!
remote_by_country = pd.crosstab(df['country'], df['remote_work'])
print("\nNumber of developers by country and remote work status:")
print(remote_by_country)


Number of developers by country and remote work status:
remote_work  hybrid  no  yes
country                     
Canada           24  25   24
Germany          32  15   25
India            38  34   42
Netherlands      14  12   14
UK               23  27   28
US               52  33   38


### EXERCISES: Section 3

**Exercise 3.1:** What is the median number of weekly hours worked? What about by experience level (using the column you created in Section 2)?


In [ ]:
# Your code here

**Exercise 3.2:** Which country has the highest average job satisfaction? Which has the lowest?


In [ ]:
# Your code here

**Exercise 3.3:** Create a cross-tabulation showing the count of developers for each combination of `experience_level` and `company_size`.


In [ ]:
# Your code here

---

## Section 4: Basic Visualization (~20 min)

### Why This Matters (From Lecture)
"A picture is worth a thousand words." Visualization was a key topic in lecture because:
- Patterns jump out visually
- You can spot outliers and distributions
- It's easier to communicate findings to others

Now you'll create the plots you saw in lecture!


### Histograms: Visualizing Numerical Distributions

A histogram shows how a numerical variable is distributed.

Reference: [matplotlib.pyplot.hist](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.hist.html)


In [ ]:
# Simple histogram of salaries
plt.figure(figsize=(10, 6))
df['annual_salary'].dropna().hist(bins=30, edgecolor='black', alpha=0.7)
plt.xlabel('Annual Salary ($)')
plt.ylabel('Frequency')
plt.title('Distribution of Developer Salaries')
plt.show()

In [ ]:
# Histogram with more control using matplotlib
plt.figure(figsize=(10, 6))
plt.hist(df['annual_salary'].dropna(), bins=25, color='steelblue', edgecolor='black', alpha=0.7)
plt.xlabel('Annual Salary ($)', fontsize=12)
plt.ylabel('Number of Developers', fontsize=12)
plt.title('Distribution of Developer Salaries', fontsize=14, fontweight='bold')
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

In [ ]:
# Histogram of years of experience
plt.figure(figsize=(10, 6))
plt.hist(df['years_experience'], bins=25, color='coral', edgecolor='black', alpha=0.7)
plt.xlabel('Years of Experience', fontsize=12)
plt.ylabel('Number of Developers', fontsize=12)
plt.title('Distribution of Developer Experience', fontsize=14, fontweight='bold')
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

### Bar Plots: Comparing Categorical Variables

Bar plots are great for comparing counts or means across categories.


In [ ]:
# Simple bar plot of language counts
language_counts = df['primary_language'].value_counts()
plt.figure(figsize=(10, 6))
language_counts.plot(kind='bar', color='steelblue', edgecolor='black', alpha=0.7)
plt.xlabel('Primary Language', fontsize=12)
plt.ylabel('Number of Developers', fontsize=12)
plt.title('Developer Count by Programming Language', fontsize=14, fontweight='bold')
plt.xticks(rotation=45)
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

In [ ]:
# Bar plot of average salary by language
avg_salary_by_lang = df.groupby('primary_language')['annual_salary'].mean().sort_values(ascending=False)
plt.figure(figsize=(10, 6))
avg_salary_by_lang.plot(kind='bar', color='mediumseagreen', edgecolor='black', alpha=0.7)
plt.xlabel('Primary Language', fontsize=12)
plt.ylabel('Average Annual Salary ($)', fontsize=12)
plt.title('Average Salary by Programming Language', fontsize=14, fontweight='bold')
plt.xticks(rotation=45)
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

### Box Plots: Comparing Distributions Across Groups

Box plots show the median, quartiles, and outliers. Great for comparing distributions!

Reference: [seaborn.boxplot](https://seaborn.pydata.org/generated/seaborn.boxplot.html)


In [ ]:
# Box plot of salary by company size
plt.figure(figsize=(10, 6))
sns.boxplot(data=df, x='company_size', y='annual_salary', palette='Set2')
plt.xlabel('Company Size', fontsize=12)
plt.ylabel('Annual Salary ($)', fontsize=12)
plt.title('Salary Distribution by Company Size', fontsize=14, fontweight='bold')
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

In [ ]:
# Box plot of salary by experience level
# First, order the experience levels logically
experience_order = ['Beginner', 'Intermediate', 'Expert']
plt.figure(figsize=(10, 6))
sns.boxplot(data=df, x='experience_level', y='annual_salary', order=experience_order, palette='husl')
plt.xlabel('Experience Level', fontsize=12)
plt.ylabel('Annual Salary ($)', fontsize=12)
plt.title('Salary Distribution by Experience Level', fontsize=14, fontweight='bold')
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

### Scatter Plots: Relationships Between Two Numerical Variables

Scatter plots show if two variables are related.


In [ ]:
# Scatter plot: experience vs salary
plt.figure(figsize=(10, 6))
plt.scatter(df['years_experience'], df['annual_salary'], alpha=0.5, s=50, color='steelblue', edgecolor='black', linewidth=0.5)
plt.xlabel('Years of Experience', fontsize=12)
plt.ylabel('Annual Salary ($)', fontsize=12)
plt.title('Relationship Between Experience and Salary', fontsize=14, fontweight='bold')
plt.grid(alpha=0.3)
plt.tight_layout()
plt.show()

In [ ]:
# Scatter plot with color coding for a third variable
languages = df['primary_language'].unique()
colors = {'Python': 'blue', 'Java': 'orange', 'JavaScript': 'green', 'C++': 'red', 'Rust': 'purple', 'Go': 'brown'}

plt.figure(figsize=(12, 6))
for lang in languages:
    mask = df['primary_language'] == lang
    plt.scatter(df[mask]['years_experience'], df[mask]['annual_salary'], 
               label=lang, alpha=0.6, s=50, color=colors.get(lang, 'gray'))

plt.xlabel('Years of Experience', fontsize=12)
plt.ylabel('Annual Salary ($)', fontsize=12)
plt.title('Experience vs Salary (colored by Language)', fontsize=14, fontweight='bold')
plt.legend(title='Language', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(alpha=0.3)
plt.tight_layout()
plt.show()

### EXERCISES: Section 4

**Exercise 4.1:** Create a histogram of job satisfaction scores. What does the distribution look like?


In [ ]:
# Your code here

**Exercise 4.2:** Create a box plot comparing job satisfaction across different remote work statuses (yes, no, hybrid). What do you observe?


In [ ]:
# Your code here

**Exercise 4.3:** Create a scatter plot showing the relationship between weekly hours worked and annual salary. Color the points by experience level. What pattern do you see?


In [ ]:
# Your code here

---

## Wrap-Up

Congratulations! You've just learned the fundamental Python skills for data analysis:

1. **Loading & Exploring Data** — `pd.read_csv()`, `.head()`, `.describe()`, column/row selection
2. **Filtering & Transforming** — Boolean masks, `.apply()`, handling missing values
3. **Data Summaries** — `.mean()`, `.median()`, `.groupby()`, `.value_counts()` (connecting to lecture concepts)
4. **Visualization** — histograms, bar plots, box plots, scatter plots

These are the **exact techniques** that data scientists use every day. Next week, you'll go deeper and learn about statistical inference!

### Key Takeaways
- Data is stored in DataFrames (rows and columns)
- You can filter, transform, and summarize data with a few lines of code
- Visualization is crucial for understanding patterns
- Missing data is common — know how to handle it
- Group operations let you compare different subsets of data
